README.md
btcd
====

[![Build Status](https://travis-ci.org/btcsuite/btcd.png?branch=master)]
(https://travis-ci.org/btcsuite/btcd)

btcd is an alternative full node bitcoin implementation written in Go (golang).

This project is currently under active development and is in a Beta state.  It
is extremely stable and has been in production use for over 6 months as of May
2014, however there are still a couple of major features we want to add before
we come out of beta.

It properly downloads, validates, and serves the block chain using the exact
rules (including bugs) for block acceptance as Bitcoin Core.  We have taken
great care to avoid btcd causing a fork to the block chain.  It passes all of
the 'official' block acceptance tests
(https://github.com/TheBlueMatt/test-scripts) as well as all of the JSON test
data in the Bitcoin Core code.

It also relays newly mined blocks, maintains a transaction pool, and relays
individual transactions that have not yet made it into a block.  It ensures all
transactions admitted to the pool follow the rules required by the block chain
and also includes the same checks which filter transactions based on
miner requirements ("standard" transactions) as Bitcoin Core.

One key difference between btcd and Bitcoin Core is that btcd does *NOT* include
wallet functionality and this was a very intentional design decision.  See the
blog entry [here](https://blog.conformal.com/btcd-not-your-moms-bitcoin-daemon)
for more details.  This means you can't actually make or receive payments
directly with btcd.  That functionality is provided by the
[btcwallet](https://github.com/conseweb/btcwallet) and
[btcgui](https://github.com/conseweb/btcgui) projects which are both under
active development.

## Requirements

[Go](http://golang.org) 1.3 or newer.

## Installation

#### Windows - MSI Available

https://github.com/conseweb/stcd/releases

#### Linux/BSD/MacOSX/POSIX - Build from Source

- Install Go according to the installation instructions here:
  http://golang.org/doc/install

- Ensure Go was installed properly and is a supported version:

```bash
$ go version
$ go env GOROOT GOPATH
```

NOTE: The `GOROOT` and `GOPATH` above must not be the same path.  It is
recommended that `GOPATH` is set to a directory in your home directory such as
`~/goprojects` to avoid write permission issues.

- Run the following command to obtain btcd, all dependencies, and install it:

```bash
$ go get -u github.com/conseweb/stcd/...
```

- btcd (and utilities) will now be installed in either ```$GOROOT/bin``` or
  ```$GOPATH/bin``` depending on your configuration.  If you did not already
  add the bin directory to your system path during Go installation, we
  recommend you do so now.

## Updating

#### Windows

Install a newer MSI

#### Linux/BSD/MacOSX/POSIX - Build from Source

- Run the following command to update btcd, all dependencies, and install it:

```bash
$ go get -u -v github.com/conseweb/stcd/...
```

## Getting Started

btcd has several configuration options avilable to tweak how it runs, but all
of the basic operations described in the intro section work with zero
configuration.

#### Windows (Installed from MSI)

Launch btcd from your Start menu.

#### Linux/BSD/POSIX/Source

```bash
$ ./btcd
````

## IRC

- irc.freenode.net
- channel #btcd
- [webchat](https://webchat.freenode.net/?channels=btcd)

## Mailing lists

- btcd: discussion of btcd and its packages.
- btcd-commits: readonly mail-out of source code changes.

To subscribe to a given list, send email to list+subscribe@opensource.conformal.com

## Issue Tracker

The [integrated github issue tracker](https://github.com/conseweb/stcd/issues)
is used for this project.

## Documentation

The documentation is a work-in-progress.  It is located in the [docs](https://github.com/conseweb/stcd/tree/master/docs) folder.

## GPG Verification Key

All official release tags are signed by Conformal so users can ensure the code
has not been tampered with and is coming from the btcsuite developers.  To
verify the signature perform the following:

- Download the public key from the Conformal website at
  https://opensource.conformal.com/GIT-GPG-KEY-conformal.txt

- Import the public key into your GPG keyring:
  ```bash
  gpg --import GIT-GPG-KEY-conformal.txt
  ```

- Verify the release tag with the following command where `TAG_NAME` is a
  placeholder for the specific tag:
  ```bash
  git tag -v TAG_NAME
  ```

## License

btcd is licensed under the [copyfree](http://copyfree.org) ISC License.


CHANGES
============================================================================
User visible changes for btcd
  A full-node bitcoin implementation written in Go
============================================================================

Changes in 0.20.0 
  - change p2p port from 8333 to 6682, rpc port from 8334 to 6684

Changes in 0.12.0 (Fri Nov 20 2015)
  - Protocol and network related changes:
    - Add a new checkpoint at block height 382320 (#555)
    - Implement BIP0065 which includes support for version 4 blocks, a new
      consensus opcode (OP_CHECKLOCKTIMEVERIFY) that enforces transaction
      lock times, and a double-threshold switchover mechanism (#535, #459,
      #455)
    - Implement BIP0111 which provides a new bloom filter service flag and
      hence provides support for protocol version 70011 (#499)
    - Add a new parameter --nopeerbloomfilters to allow disabling bloom
      filter support (#499)
    - Reject non-canonically encoded variable length integers (#507)
    - Add mainnet peer discovery DNS seed (seed.bitcoin.jonasschnelli.ch)
      (#496)
    - Correct reconnect handling for persistent peers (#463, #464)
    - Ignore requests for block headers if not fully synced (#444)
    - Add CLI support for specifying the zone id on IPv6 addresses (#538)
    - Fix a couple of issues where the initial block sync could stall (#518,
      #229, #486)
    - Fix an issue which prevented the --onion option from working as
      intended (#446)
  - Transaction relay (memory pool) changes:
    - Require transactions to only include signatures encoded with the
    canonical 'low-s' encoding (#512)
    - Add a new parameter --minrelaytxfee to allow the minimum transaction
      fee in BTC/kB to be overridden (#520)
    - Retain memory pool transactions when they redeem another one that is
      removed when a block is accepted (#539)
    - Do not send reject messages for a transaction if it is valid but
      causes an orphan transaction which depends on it to be determined
      as invalid (#546)
    - Refrain from attempting to add orphans to the memory pool multiple
      times when the transaction they redeem is added (#551)
    - Modify minimum transaction fee calculations to scale based on bytes
      instead of full kilobyte boundaries (#521, #537)
  - Implement signature cache:
    - Provides a limited memory cache of validated signatures which is a
      huge optimization when verifying blocks for transactions that are
      already in the memory pool (#506)
    - Add a new parameter '--sigcachemaxsize' which allows the size of the
      new cache to be manually changed if desired (#506)
  - Mining support changes:
    - Notify getblocktemplate long polling clients when a block is pushed
      via submitblock (#488)
    - Speed up getblocktemplate by making use of the new signature cache
      (#506)
  - RPC changes:
    - Implement getmempoolinfo command (#453)
    - Implement getblockheader command (#461)
    - Modify createrawtransaction command to accept a new optional parameter
      'locktime' (#529)
    - Modify listunspent result to include the 'spendable' field (#440)
    - Modify getinfo command to include 'errors' field (#511)
    - Add timestamps to blockconnected and blockdisconnected notifications
      (#450)
    - Several modifications to searchrawtranscations command:
      - Accept a new optional parameter 'vinextra' which causes the results
        to include information about the outputs referenced by a transaction's
        inputs (#485, #487)
      - Skip entries in the mempool too (#495)
      - Accept a new optional parameter 'reverse' to return the results in
        reverse order (most recent to oldest) (#497)
      - Accept a new optional parameter 'filteraddrs' which causes the
        results to only include inputs and outputs which involve the
        provided addresses (#516)
    - Change the notification order to notify clients about mined
      transactions (recvtx, redeemingtx) before the blockconnected
      notification (#449)
    - Update verifymessage RPC to use the standard algorithm so it is
      compatible with other implementations (#515)
    - Improve ping statistics by pinging on an interval (#517)
  - Websocket changes:
    - Implement session command which returns a per-session unique id (#500,
      #503)
  - btcctl utility changes:
    - Add getmempoolinfo command (#453)
    - Add getblockheader command (#461)
    - Add getwalletinfo command (#471)
  - Notable developer-related package changes:
    - Introduce a new peer package which acts a common base for creating and
      concurrently managing bitcoin network peers (#445)
    - Various cleanup of the new peer package (#528, #531, #524, #534,
      #549)
    - Blocks heights now consistently use int32 everywhere (#481)
    - The BlockHeader type in the wire package now provides the BtcDecode
      and BtcEncode methods (#467)
    - Update wire package to recognize BIP0064 (getutxo) service bit (#489)
    - Export LockTimeThreshold constant from txscript package (#454)
    - Export MaxDataCarrierSize constant from txscript package (#466)
    - Provide new IsUnspendable function from the txscript package (#478)
    - Export variable length string functions from the wire package (#514)
    - Export DNS Seeds for each network from the chaincfg package (#544)
    - Preliminary work towards separating the memory pool into a separate
      package (#525, #548)
  - Misc changes:
    - Various documentation updates (#442, #462, #465, #460, #470, #473,
      #505, #530, #545)
    - Add installation instructions for gentoo (#542)
    - Ensure an error is shown if OS limits can't be set at startup (#498)
    - Tighten the standardness checks for multisig scripts (#526)
    - Test coverage improvement (#468, #494, #527, #543, #550)
    - Several optimizations (#457, #474, #475, #476, #508, #509)
    - Minor code cleanup and refactoring (#472, #479, #482, #519, #540)

doc.go
// Copyright (c) 2013-2014 The btcsuite developers
// Use of this source code is governed by an ISC
// license that can be found in the LICENSE file.

/*
btcd is a full-node bitcoin implementation written in Go.

The default options are sane for most users.  This means btcd will work 'out of
the box' for most users.  However, there are also a wide variety of flags that
can be used to control it.

The following section provides a usage overview which enumerates the flags.  An
interesting point to note is that the long form of all of these options
(except -C) can be specified in a configuration file that is automatically
parsed when btcd starts up.  By default, the configuration file is located at
~/.btcd/btcd.conf on POSIX-style operating systems and %LOCALAPPDATA%\btcd\btcd.conf
on Windows.  The -C (--configfile) flag, as shown below, can be used to override
this location.

Usage:
  btcd [OPTIONS]

Application Options:
  -V, --version             Display version information and exit
  -C, --configfile=         Path to configuration file
  -b, --datadir=            Directory to store data
      --logdir=             Directory to log output.
  -a, --addpeer=            Add a peer to connect with at startup
      --connect=            Connect only to the specified peers at startup
      --nolisten            Disable listening for incoming connections -- NOTE:
                            Listening is automatically disabled if the --connect
                            or --proxy options are used without also specifying
                            listen interfaces via --listen
      --listen=             Add an interface/port to listen for connections
                            (default all interfaces port: 6682, testnet: 16682)
      --maxpeers=           Max number of inbound and outbound peers (125)
      --banduration=        How long to ban misbehaving peers.  Valid time units
                            are {s, m, h}.  Minimum 1 second (24h0m0s)
  -u, --rpcuser=            Username for RPC connections
  -P, --rpcpass=            Password for RPC connections
      --rpclimituser=       Username for limited RPC connections
      --rpclimitpass=       Password for limited RPC connections
      --rpclisten=          Add an interface/port to listen for RPC connections
                            (default port: 6684, testnet: 16684)
      --rpccert=            File containing the certificate file
      --rpckey=             File containing the certificate key
      --rpcmaxclients=      Max number of RPC clients for standard connections
                            (10)
      --rpcmaxwebsockets=   Max number of RPC websocket connections (25)
      --norpc               Disable built-in RPC server -- NOTE: The RPC server
                            is disabled by default if no rpcuser/rpcpass or
                            rpclimituser/rpclimitpass is specified
      --notls               Disable TLS for the RPC server -- NOTE: This is only
                            allowed if the RPC server is bound to localhost
      --nodnsseed           Disable DNS seeding for peers
      --externalip=         Add an ip to the list of local addresses we claim to
                            listen on to peers
      --proxy=              Connect via SOCKS5 proxy (eg. 127.0.0.1:9050)
      --proxyuser=          Username for proxy server
      --proxypass=          Password for proxy server
      --onion=              Connect to tor hidden services via SOCKS5 proxy
                            (eg. 127.0.0.1:9050)
      --onionuser=          Username for onion proxy server
      --onionpass=          Password for onion proxy server
      --noonion             Disable connecting to tor hidden services
      --torisolation        Enable Tor stream isolation by randomizing user
                            credentials for each connection.
      --testnet             Use the test network
      --regtest             Use the regression test network
      --simnet              Use the simulation test network
      --nocheckpoints       Disable built-in checkpoints.  Don't do this unless
                            you know what you're doing.
      --dbtype=             Database backend to use for the Block Chain
                            (leveldb)
      --profile=            Enable HTTP profiling on given port -- NOTE port
                            must be between 1024 and 65536
      --cpuprofile=         Write CPU profile to the specified file
  -d, --debuglevel=         Logging level for all subsystems {trace, debug,
                            info, warn, error, critical} -- You may also specify
                            <subsystem>=<level>,<subsystem2>=<level>,... to set
                            the log level for individual subsystems -- Use show
                            to list available subsystems (info)
      --upnp                Use UPnP to map our listening port outside of NAT
      --minrelaytxfee=      The minimum transaction fee in BTC/kB to be
                            considered a non-zero fee.
      --limitfreerelay=     Limit relay of transactions with no transaction fee
                            to the given amount in thousands of bytes per
                            minute (15)
      --norelaypriority     Do not require free or low-fee transactions to have
                            high priority for relaying
      --maxorphantx=        Max number of orphan transactions to keep in memory
                            (1000)
      --generate            Generate (mine) bitcoins using the CPU
      --miningaddr=         Add the specified payment address to the list of
                            addresses to use for generated blocks -- At least
                            one address is required if the generate option is
                            set
      --blockminsize=       Mininum block size in bytes to be used when creating
                            a block
      --blockmaxsize=       Maximum block size in bytes to be used when creating
                            a block (750000)
      --blockprioritysize=  Size in bytes for high-priority/low-fee transactions
                            when creating a block (50000)
      --getworkkey=         DEPRECATED -- Use the --miningaddr option instead
      --addrindex           Build and maintain a full address index. Currently
                            only supported by leveldb.
      --dropaddrindex       Deletes the address-based transaction index from the
                            database on start up, and the exits.
      --nopeerbloomfilters  Disable bloom filtering support.
      --sigcachemaxsize=    The maximum number of entries in the signature
                            verification cache.

Help Options:
  -h, --help           Show this help message

*/
package main
